File: browser-engine/src/utils/Fingerprint.ts
/**
 * 浏览器指纹管理工具
 *
 * 生成随机化的浏览器指纹，用于反检测
 * 包括 UserAgent、视口、WebGL、Canvas 等参数
 */

export interface FingerprintConfig {
  userAgent: string;
  viewport: { width: number; height: number };
  platform: string;
  language: string;
  webglVendor: string;
  webglRenderer: string;
  canvasNoise: number;
  audioNoise: number;
  screenResolution: { width: number; height: number };
  colorDepth: number;
  deviceMemory: number;
  hardwareConcurrency: number;
}

// 常用 UserAgent 列表（Chrome 最新版本）
const USER_AGENTS = [
  'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36',
  'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36',
  'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36',
  'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36',
  'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36',
  'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36',
  'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36',
  'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36',
];

// 常用视口尺寸
const VIEWPORTS = [
  { width: 1920, height: 1080 },
  { width: 1366, height: 768 },
  { width: 1536, height: 864 },
  { width: 1440, height: 900 },
  { width: 1280, height: 720 },
  { width: 1600, height: 900 },
  { width: 2560, height: 1440 },
  { width: 1920, height: 1200 },
];

// WebGL 厂商和渲染器
const WEBGL_VENDORS = [
  'Google Inc. (NVIDIA)',
  'Google Inc. (Intel)',
  'Google Inc. (AMD)',
  'Intel Inc.',
  'NVIDIA Corporation',
  'ATI Technologies Inc.',
];

const WEBGL_RENDERERS = [
  'ANGLE (NVIDIA, NVIDIA GeForce RTX 3060 Direct3D11 vs_5_0 ps_5_0)',
  'ANGLE (NVIDIA, NVIDIA GeForce GTX 1660 Ti Direct3D11 vs_5_0 ps_5_0)',
  'ANGLE (Intel, Intel(R) UHD Graphics 630 Direct3D11 vs_5_0 ps_5_0)',
  'ANGLE (Intel, Intel(R) Iris(R) Xe Graphics Direct3D11 vs_5_0 ps_5_0)',
  'ANGLE (AMD, AMD Radeon RX 580 Direct3D11 vs_5_0 ps_5_0)',
  'ANGLE (AMD, AMD Radeon RX 6700 XT Direct3D11 vs_5_0 ps_5_0)',
];

// 平台标识
const PLATFORMS = ['Win32', 'MacIntel', 'Linux x86_64'];

// 语言
const LANGUAGES = ['zh-CN', 'en-US', 'zh-TW', 'en-GB', 'ja-JP'];

/**
 * 随机选择数组中的一个元素
 */
function randomPick<T>(arr: T[]): T {
  return arr[Math.floor(Math.random() * arr.length)];
}

/**
 * 生成随机整数（包含 min 和 max）
 */
function randomInt(min: number, max: number): number {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

/**
 * 生成完整的随机浏览器指纹
 */
export function generateFingerprint(): FingerprintConfig {
  const viewport = getViewport();

  return {
    userAgent: getUserAgent(),
    viewport,
    platform: randomPick(PLATFORMS),
    language: randomPick(LANGUAGES),
    webglVendor: randomPick(WEBGL_VENDORS),
    webglRenderer: randomPick(WEBGL_RENDERERS),
    canvasNoise: Math.random() * 0.01,  // 0-0.01 的微小噪声
    audioNoise: Math.random() * 0.001,   // 0-0.001 的音频噪声
    screenResolution: {
      width: viewport.width + randomInt(0, 100),
      height: viewport.height + randomInt(40, 120), // 加上任务栏高度
    },
    colorDepth: randomPick([24, 32]),
    deviceMemory: randomPick([4, 8, 16, 32]),
    hardwareConcurrency: randomPick([4, 6, 8, 12, 16]),
  };
}

/**
 * 获取随机 UserAgent
 */
export function getUserAgent(): string {
  return randomPick(USER_AGENTS);
}

/**
 * 获取随机视口大小
 */
export function getViewport(): { width: number; height: number } {
  return { ...randomPick(VIEWPORTS) };
}

/**
 * 获取 WebGL 参数
 */
export function getWebGLParams(): { vendor: string; renderer: string } {
  return {
    vendor: randomPick(WEBGL_VENDORS),
    renderer: randomPick(WEBGL_RENDERERS),
  };
}

/**
 * 获取 Canvas 噪声参数
 */
export function getCanvasNoise(): number {
  return Math.random() * 0.01;
}

/**
 * 生成随机的 WebGL 哈希（用于 Canvas 指纹伪造）
 */
export function generateWebGLHash(): string {
  const chars = '0123456789abcdef';
  let hash = '';
  for (let i = 0; i < 32; i++) {
    hash += chars[Math.floor(Math.random() * chars.length)];
  }
  return hash;
}

File: desktop-companion/fingerprint.py
"""
fingerprint.py — 每账号独立浏览器指纹生成。

为每个账号生成确定性但唯一的指纹，避免平台风控检测到
多个账号来自"同一台设备"。
"""

import hashlib
import json
import random
from pathlib import Path

# ── 真实 GPU 配置池（vendor, renderer）──
_GPU_PROFILES = [
    ('Google Inc. (Intel)', 'ANGLE (Intel, Intel(R) UHD Graphics 630 Direct3D11 vs_5_0 ps_5_0, D3D11)'),
    ('Google Inc. (Intel)', 'ANGLE (Intel, Intel(R) UHD Graphics 770 Direct3D11 vs_5_0 ps_5_0, D3D11)'),
    ('Google Inc. (Intel)', 'ANGLE (Intel, Intel(R) Iris(R) Xe Graphics Direct3D11 vs_5_0 ps_5_0, D3D11)'),
    ('Google Inc. (Intel)', 'ANGLE (Intel, Intel(R) HD Graphics 530 Direct3D11 vs_5_0 ps_5_0, D3D11)'),
    ('Google Inc. (NVIDIA)', 'ANGLE (NVIDIA, NVIDIA GeForce GTX 1660 Direct3D11 vs_5_0 ps_5_0, D3D11)'),
    ('Google Inc. (NVIDIA)', 'ANGLE (NVIDIA, NVIDIA GeForce RTX 3060 Direct3D11 vs_5_0 ps_5_0, D3D11)'),
    ('Google Inc. (NVIDIA)', 'ANGLE (NVIDIA, NVIDIA GeForce RTX 4060 Direct3D11 vs_5_0 ps_5_0, D3D11)'),
    ('Google Inc. (NVIDIA)', 'ANGLE (NVIDIA, NVIDIA GeForce GTX 1050 Ti Direct3D11 vs_5_0 ps_5_0, D3D11)'),
    ('Google Inc. (AMD)', 'ANGLE (AMD, AMD Radeon RX 580 Direct3D11 vs_5_0 ps_5_0, D3D11)'),
    ('Google Inc. (AMD)', 'ANGLE (AMD, AMD Radeon RX 6600 XT Direct3D11 vs_5_0 ps_5_0, D3D11)'),
    ('Google Inc. (AMD)', 'ANGLE (AMD, AMD Radeon RX 7600 Direct3D11 vs_5_0 ps_5_0, D3D11)'),
]

# ── Chrome 版本池（匹配 Playwright Chromium 大版本区间）──
_CHROME_VERSIONS = ['126', '127', '128', '129', '130', '131', '132', '133']

# ── 硬件参数池 ──
_HARDWARE_CONCURRENCY = [4, 6, 8, 12, 16]
_DEVICE_MEMORY = [4, 8, 16]

# ── 屏幕分辨率池 ──
_SCREEN_SIZES = [
    (1920, 1080), (2560, 1440), (1366, 768), (1440, 900), (1536, 864),
    (1680, 1050), (1280, 720),
]

# ── Canvas 噪声种子范围 ──
_CANVAS_NOISE_RANGE = (-3, 3)


def generate_fingerprint(seed_str: str) -> dict:
    """从种子字符串（account_id 或随机 UUID）生成确定性指纹。

    Args:
        seed_str: 账号ID或任意唯一字符串

    Returns:
        指纹字典，包含 WebGL、UA、硬件参数等
    """
    seed = int(hashlib.md5(seed_str.encode('utf-8')).hexdigest(), 16)
    rng = random.Random(seed)

    gpu_vendor, gpu_renderer = rng.choice(_GPU_PROFILES)
    chrome_ver = rng.choice(_CHROME_VERSIONS)
    hw_conc = rng.choice(_HARDWARE_CONCURRENCY)
    dev_mem = rng.choice(_DEVICE_MEMORY)
    screen_w, screen_h = rng.choice(_SCREEN_SIZES)
    canvas_noise = rng.randint(_CANVAS_NOISE_RANGE[0], _CANVAS_NOISE_RANGE[1])

    return {
        'webgl_vendor': gpu_vendor,
        'webgl_renderer': gpu_renderer,
        'chrome_version': chrome_ver,
        'user_agent': (
            f'Mozilla/5.0 (Windows NT 10.0; Win64; x64) '
            f'AppleWebKit/537.36 (KHTML, like Gecko) '
            f'Chrome/{chrome_ver}.0.0.0 Safari/537.36'
        ),
        'hardware_concurrency': hw_conc,
        'device_memory': dev_mem,
        'screen_width': screen_w,
        'screen_height': screen_h,
        'canvas_noise': canvas_noise,
    }


def save_fingerprint(profile_dir: Path, fingerprint: dict) -> None:
    """保存指纹到 profile 目录。"""
    fp_path = Path(profile_dir) / 'fingerprint.json'
    fp_path.parent.mkdir(parents=True, exist_ok=True)
    fp_path.write_text(json.dumps(fingerprint, ensure_ascii=False), encoding='utf-8')


def load_fingerprint(profile_dir: Path, account_id: str = '') -> dict:
    """从 profile 目录加载指纹，不存在则从 account_id 生成。"""
    fp_path = Path(profile_dir) / 'fingerprint.json'
    if fp_path.exists():
        try:
            return json.loads(fp_path.read_text('utf-8'))
        except Exception:
            pass
    # Fallback: 从 account_id 生成
    if account_id:
        fp = generate_fingerprint(account_id)
        try:
            save_fingerprint(profile_dir, fp)
        except Exception:
            pass
        return fp
    # Last resort: default
    return generate_fingerprint('default_seed_000')
